### PR TITLE
sgx-sdk: improve derivation

### PR DESCRIPTION
**Pull request by veehaitch**
###### Motivation for this change

Follow up for #126990

Feel free to squash on merge.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veehaitch** [comment]
Restructured the directory hierarchy of the derivations output. I also added some symlinks to keep it compatible to the structure created by the Intel SDK installer.

```shell
# tree -L 1 ./result
./result/
├── bin
├── buildenv.mk -> share/bin/buildenv.mk
├── environment -> share/bin/environment
├── include
├── lib
├── lib64 -> lib/
├── pkgconfig -> lib/pkgconfig/
├── sdk_libs
├── sgxsdk -> .
└── share

8 directories, 2 files
```

<details>
  <summary>tree -L 2 ./result</summary>

```shell
./result/
├── bin
│   ├── sgx_config_cpusvn
│   ├── sgx_edger8r
│   ├── sgx_encrypt
│   ├── sgx-gdb
│   ├── sgx_sign
│   └── x64
├── buildenv.mk -> share/bin/buildenv.mk
├── environment -> share/bin/environment
├── include
│   ├── ipp
│   ├── libcxx
│   ├── sgx_attributes.h
│   ├── sgx_capable.h
│   ├── sgx_cpuid.h
│   ├── sgx_dcap_tvl.edl
│   ├── sgx_dcap_tvl.h
│   ├── sgx_defs.h
│   ├── sgx_dh.h
│   ├── sgx_ecp_types.h
│   ├── sgx_edger8r.h
│   ├── sgx_eid.h
│   ├── sgx_enclave_common.h
│   ├── sgx_error.h
│   ├── sgx.h
│   ├── sgx_intrin.h
│   ├── sgx_key_exchange.h
│   ├── sgx_key.h
│   ├── sgx_lfence.h
│   ├── sgx_pce.h
│   ├── sgx_pcl_guid.h
│   ├── sgx_pthread.edl
│   ├── sgx_ql_lib_common.h
│   ├── sgx_ql_quote.h
│   ├── sgx_quote_3.h
│   ├── sgx_quote.h
│   ├── sgx_qve_header.h
│   ├── sgx_report.h
│   ├── sgx_rsrv_mem_mngr.h
│   ├── sgx_secure_align_api.h
│   ├── sgx_secure_align.h
│   ├── sgx_spinlock.h
│   ├── sgx_tcrypto.h
│   ├── sgx_thread.h
│   ├── sgx_tkey_exchange.edl
│   ├── sgx_tkey_exchange.h
│   ├── sgx_tprotected_fs.edl
│   ├── sgx_tprotected_fs.h
│   ├── sgx_trts_exception.h
│   ├── sgx_trts.h
│   ├── sgx_tseal.h
│   ├── sgx_tstdc.edl
│   ├── sgx_tswitchless.edl
│   ├── sgx_uae_epid.h
│   ├── sgx_uae_launch.h
│   ├── sgx_uae_quote_ex.h
│   ├── sgx_uae_service.h
│   ├── sgx_ukey_exchange.h
│   ├── sgx_urts.h
│   ├── sgx_uswitchless.h
│   ├── sgx_utils.h
│   ├── stdc++
│   └── tlibc
├── lib
│   ├── cve_2020_0551_cf
│   ├── cve_2020_0551_load
│   ├── gdb-sgx-plugin
│   ├── libc++_Changes_SGX.txt
│   ├── libsgx_capable.a
│   ├── libsgx_capable.so
│   ├── libsgx_dcap_tvl.a
│   ├── libsgx_epid_sim.so
│   ├── libsgx_epid.so
│   ├── libsgx_launch_sim.so
│   ├── libsgx_launch.so
│   ├── libsgx_omp.a
│   ├── libsgx_pcl.a
│   ├── libsgx_pclsim.a
│   ├── libsgx_pthread.a
│   ├── libsgx_ptrace.so
│   ├── libsgx_quote_ex_sim.so
│   ├── libsgx_quote_ex.so
│   ├── libsgx_tcmalloc.a
│   ├── libsgx_tcrypto.a
│   ├── libsgx_tcxx.a
│   ├── libsgx_tkey_exchange.a
│   ├── libsgx_tprotected_fs.a
│   ├── libsgx_trts.a
│   ├── libsgx_trts_sim.a
│   ├── libsgx_tservice.a
│   ├── libsgx_tservice_sim.a
│   ├── libsgx_tstdc.a
│   ├── libsgx_tswitchless.a
│   ├── libsgx_uae_service_sim.so
│   ├── libsgx_uae_service.so
│   ├── libsgx_ukey_exchange.a
│   ├── libsgx_uprotected_fs.a
│   ├── libsgx_urts_sim.so
│   ├── libsgx_urts.so
│   ├── libsgx_uswitchless.a
│   └── pkgconfig
├── lib64 -> lib/
├── pkgconfig -> lib/pkgconfig/
├── sdk_libs
│   ├── libsgx_epid_sim.so -> ../lib/libsgx_epid_sim.so
│   ├── libsgx_launch_sim.so -> ../lib/libsgx_launch_sim.so
│   ├── libsgx_quote_ex_sim.so -> ../lib/libsgx_quote_ex_sim.so
│   ├── libsgx_uae_service_sim.so -> ../lib/libsgx_uae_service_sim.so
│   └── libsgx_urts_sim.so -> ../lib/libsgx_urts_sim.so
├── sgxsdk -> .
└── share
    ├── bin
    ├── licenses
    └── SampleCode

20 directories, 94 files

```
</details>

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:128]: I wasn't really sure about the purpose of that part. The build and check phase run fine without it. Maybe I'm missing something though.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:141]: I think it's fine to do the regular fixup routines. As opposed to the AESM service, there are no prebuilt and signed binaries involved, are there?

**Comment by veehaitch** [comment]
Result of `nixpkgs-review pr 146058` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sgx-sdk</li>
  </ul>
</details>

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:141]:
```suggestion
    mkdir -p $out/bin
    pushd $out
```

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:196]:
```suggestion
    for path in "$out/share/bin/environment" "$out/bin/sgx-gdb"; do
```

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:233]:
```suggestion
    TESTDIR=$(mktemp -d)
```

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:238]: Missing -j$NIX_BUILD_CORES

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:244]: Please either use cd or popd.

**Comment by veehaitch** [comment]
Result of `nixpkgs-review pr 146058` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sgx-sdk</li>
  </ul>
</details>

**Comment by veehaitch** [comment]
Instead of building the samples as part of the `installCheckPhase`, I created individual derivations for each sample. Those derivations use `sgx-sdk` as a build input. The setup hook makes sure the environment of the downstream derivation is initialized appropriately. 

I added the samples to `passthru.tests` of the `sgx-sdk` derivation. @GrahamcOfBorg runs those tests automatically. To run them manually, execute the following command from the root of your `nixpkgs` checkout: `nix-build -A sgx-sdk.tests`.

I guess that's it. I'd suggest we do the update to `2.15` in another PR following this one. Would appreciate it if you, @sbellem and @arcz, could take a look.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (arcz) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:141]: right, tbh. I'm not sure why it was needed

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (arcz) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:128]: I double checked and the build fails for me after removing `SHELL=$SHELL`. Maybe the default build script sets that up so after you moved things around it's no longer needed?

Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:128]: Yes, you're right:

https://github.com/NixOS/nixpkgs/blob/f8f45750307cc93c6925f485c374818a51c8c5c6/pkgs/stdenv/generic/setup.sh#L1064-L1069

**Comment by sbellem** [comment]
> I guess that's it. I'd suggest we do the update to `2.15` in another PR following this one. Would appreciate it if you, @sbellem and @arcz, could take a look.

Thanks @veehaitch! I will be reviewing the PR tomorrow (Wednesday).

**Comment by veehaitch** [comment]
Hey @sbellem, did you already have a chance to take a look?

**Comment by sbellem** [comment]
> Hey @sbellem, did you already have a chance to take a look?

Sorry, I have been delayed. I just looked at your latest commit, and that looks good. I am looking at the rest of the PR, today.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (sbellem) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:51]: Cool! I did not know that the missing header file was under this `external/epid-sdk`.

Just wish to note here that doing `diff` between this `external/epid-sdk/ .../sgx_ippcp.h` and the one that is packaged in "Prebuilt Optimized Libraries Binaries" at https://01.org/intel-softwareguard-extensions/downloads/intel-sgx-linux-2.14-release, does not match. After inspection, it can be seen that it's only the formatting and comments that differ. 

However, it appears that this `external/epid-sdk/ext/ipp/` is not kept up-to-date. It has not been updated since release 2.13.3, and in the case of 2.14, it may just happen that we're lucky that the header file (`sgx_ippcp.h`) did not change, but it does in 2.15. I do not know what is the significance of these changes, but just wish to point out that this approach may be somewhat fragile, as future updates to the derivation for newer sgx sdk versions, will require a different approach if the source repo (intel/linux-sgx) is not updated to match what is published/packaged by Intel.

It would be best to see with Intel, to make sure that they keep the header file in sync with what they publish for download. Opened and issue: https://github.com/intel/linux-sgx/issues/759.

In the meantime, it seems to me that sticking to the patch/commit/PR would be simpler, and easier to double-check if needed, but I don't mind if you prefer the newer proposed approach. :)

**Review by sbellem — COMMENTED** [review]
It looks good to me overall! Thanks for helping improving!

**Comment by sbellem** [comment]
@veehaitch I added 2 comments. The first about the commit hash versus the tag appears to me quite important. If we somehow decide to stick with using the tag I'd like to better understand why it could not be problematic for reproducible builds.

I have yet to try/test the derivation, and should do so tomorrow.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:34]: Thanks for being so considerate about reproducibility! You can, however, rest assured—nothing can go wrong here with Nix: the `fetchFromGitHub` call results in a fixed output derivation which has to match the given `sha256`. If someone updates the tag to point to a different commit, two things can happen:

1. Fetching the source fails with a hash mismatch.
2. If the given hash points to a derivation which is already in the store of the executing machine (or in any configured binary cache), it will silently use the original version before the tag was updated in the upstream repository. This easily leads to confusion when updating a `rev` without updating the `sha256` (or at least set it to a non-existing path, e.g., `lib.fakeSha256`).

Overall, what really matters is the hash. If the hash corresponds to the version you want to use, nothing can go wrong.

Still, I cannot confirm that the tag `sgx_2.14` points to a different commit. It's exactly the same:

```shell
# git rev-list -n 1 sgx_2.14
0cea078f17a24fb807e706409972d77f7a958db9
```

https://github.com/NixOS/nixpkgs/blob/a0ab6d84ff85a81feb1081fb46acfdc644feb7d6/pkgs/os-specific/linux/sgx-sdk/default.nix#L33

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:51]: Yes, I also inspected the diff of your patch and the version from `epid-sdk` and concluded that they match.

In the end, it doesn't really matter if we add the missing file (your approach) or if we copy/link it (my approach). Ultimately, both approaches are a patch and we have to keep it in sync with the corresponding versions. Therefore, it makes a lot of sense that you already opened upstream issues.

I'm not against using a patch. For the current version it just made sense to use the version we already have in the source derivation. I'm fine with going back to a patch but I'd prefer adding the patch file (i.e., check it in).

Maybe it's most idiomatic to use `fetchurl` to fetch the `sgx_ippcp.h` file from the original source were you created the patch from?

Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix]: > Maybe it's most idiomatic to use `fetchurl` to fetch the `sgx_ippcp.h` file from the original source were you created the patch from?

See https://github.com/NixOS/nixpkgs/pull/146058/commits/406fb9b14db226fed9b49d6e95b0373683b2a586.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (sbellem) [reply pkgs/os-specific/linux/sgx-sdk/default.nix]: :smile: reproducible builds was the main concern that drove me to `nix` :smile:

Ok, I see, sorry, I overlooked the hash! Totally fine then. As for the commit, I got misled by looking up the tag on GitHub at https://github.com/intel/linux-sgx/commits/sgx_2.14. When checking the tag, you're right, it points to `0cea078f17a24fb807e706409972d77f7a958db9`.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (sbellem) [reply pkgs/os-specific/linux/sgx-sdk/default.nix]: Oh thanks for going through the effort for this different approach. I think it makes it easier to quickly double check that the header is the expected one, but perhaps that is just a personal preference. Either way, I like the new approach, and agree with you that either way it is temporary as we would prefer upstream to take proper care of this. :+1:

**Comment by sbellem** [comment]
@veehaitch Ran `nix-build -A sgx-sdk.tests` and everything looks good to me.

If you don't mind I'll also try the derivation with some toy enclaves that I use. I expect to be done with this by Thursday at the latest.

Very nice work!

**Comment by veehaitch** [comment]
> If you don't mind I'll also try the derivation with some toy enclaves that I use. I expect to be done with this by Thursday at the latest.

Sure, go ahead.

I just tested this PR with the remote attestation sample with `SGX_MODE=HW` using SGX AESM from [`sgx-psw`](https://github.com/veehaitch/nixpkgs/compare/sgx-sdk-improve-drv...sgx-psw) (going to open a PR as soon as this one is merged). It works fine 🙂

**Comment by veehaitch** [comment]
Any final thoughts, @sbellem @arcz? If not, please approve this PR to get it merged 🙂

**Review by sbellem — APPROVED** [review]

**Comment by veehaitch** [comment]
Result of `nixpkgs-review pr 146058` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sgx-sdk</li>
  </ul>
</details>

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:86]: How is ncurses used in the build?

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/sgx-sdk/default.nix]: Does this run on amd cpus as well?

File: pkgs/os-specific/linux/sgx-sdk/samples.nix
Comment (Mic92) [inline pkgs/os-specific/linux/sgx-sdk/samples.nix:17]: Does this run on amd cpus or on intel CPUs without SGX as well?

Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/samples.nix:17]: Yes, this should work. I didn't test it though as I don't have a machine with an AMD CPU available.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:86]: Very good question. I just built the derivation without it and it turns out: it is not required. Going to remove it.

File: pkgs/os-specific/linux/sgx-sdk/samples.nix
Comment (Mic92) [inline pkgs/os-specific/linux/sgx-sdk/samples.nix:15]:
```suggestion
    enableParallelBuilding = true;   
```

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/sgx-sdk/default.nix:143]:
```suggestion
  ];
  
  enableParallelBuilding = true;
```

File: pkgs/os-specific/linux/sgx-sdk/samples.nix
Comment (Mic92) [reply pkgs/os-specific/linux/sgx-sdk/samples.nix:17]: Works on AMD EPYC.

File: pkgs/os-specific/linux/sgx-sdk/samples.nix
Comment (Mic92) [reply pkgs/os-specific/linux/sgx-sdk/samples.nix:15]: This saved quite some time when testing.

File: pkgs/os-specific/linux/sgx-sdk/default.nix
Comment (veehaitch) [reply pkgs/os-specific/linux/sgx-sdk/default.nix:143]: In general, this is a good idea. Apparently, however, dependencies are not properly declared, hence, parallel building fails sometimes. Properly we should even consider `enableParallelBuilding = false` with a comment.

**Review by Mic92 — APPROVED** [review]
will merge once my nixpkgs-review went through.

**Comment by Mic92** [comment]
merged in https://github.com/NixOS/nixpkgs/pull/148584

**Pull request closed without merging**
